File: C2/PDOO/Practicas/DeepSpaceJava/src/deepspace/CardDeck.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Random;

/**
 *
 * @author Profesor
 * @param <T>
 */
class CardDeck<T> {    
    private ArrayList<T> cards=new ArrayList();
    private boolean ready;
    private int count;
    
    
    CardDeck() {
        ready=false;
        count=0;
    }
    
    public void add(T t) {
        if (!ready)
            cards.add(t);
    }
    
    public T next() {
        if (!ready) {
            ready=true;
            shuffle();
        }

        T card=cards.remove(0);
        cards.add(card);
        
        count++;
        if (count==cards.size()) {
            shuffle();
            count=0;
        }
        return card;
    }
    
    private void shuffle() {
        Collections.shuffle(cards);
    }
    
    boolean justShuffled() {
        return (count==0);
    }
}

File: C2/PDOO/Practicas/DeepSpaceRuby/lib/CardDeck.rb
#encoding:utf-8

module Deepspace

# 2.3.17 - Translation from Java
# @author Profesor
    
class CardDeck 
  
  def initialize()
    @cards=Array.new()
    @ready=false
    @count=0
  end
  
  public
  
  def add (t)
    if !@ready then
      @cards << t
    end
  end
  
  def next ()
    if (!@ready) then
      @ready=true;
      shuffle();
    end
      
    card=@cards.delete_at(0);
    @cards.push(card);
    
    @count+=1
    if (@count==@cards.size()) then
      shuffle();
      @count=0;
    end
      
    return card.class.newCopy(card)
  end
  
  def justShuffled()
    return (@count==0)
  end

  private
  
  def shuffle()
    @cards.shuffle!
  end
  
  def to_s
    getUIversion().to_s
  end
  
end # class

end # module

if $0 == __FILE__ then
  class TestCard
    attr_reader :a
    def initialize (a)
      @a=a
    end
  end
  
  test=Deepspace::CardDeck.new
  test.add(TestCard.new(1))
  test.add(TestCard.new(2))
  test.add(TestCard.new(3))
  test.add(TestCard.new(4))
  test.add(TestCard.new(5))
  for i in 0..15 do
    puts test.next.a
  end
end
